### ISSUE TITLE
OData sends literals instead of parameters to the Queryable.

### ISSUE THREAD

**Issue by mburbea** [OData/WebApi#1843]
Using aspnetcore 2.2 and odata 7.1

Create a table like so::
```sql
create table Issue1843
(
      Enum INT not null,
      DateTime2 DATETIME2 not null
)
```
Create a controller for the following class
```cs
public enum MyEnum{ F0}

public class Issue1843
{
     [Key]
     public MyEnum Enum { get; set; }
     public DateTime DateTime2{ get;set; }
}
public class Issue1843Controller : ODataController
{
    private readonly Issue1843DbContext _context;

    public Issue1843Controller(Issue1843DbContext context)
    {
        _context = context;
    }

    [EnableQuery]
    public IActionResult Get()
    {
        return Ok(_context.Issue1843);
    }
}
public class Issue1843DbContext : DbContext
{
       public DbSet<Issue1843> Issue1843 { get;set; }
}
```

Now using your browser hit this end point with the following filter:
`/Issue1843?$filter=Enum eq 'F0' and DateTime2 eq 2019-07-02T00:00:00Z`

When you see the query generated it will be the following:
```sql
SELECT [$it].[Enum], [$it].[DateTime2]
FROM [Issue1843] AS [$it]
WHERE ([$it].[Enum] = 0) AND ([$it].[DateTime2] = '2019-07-02T00:00:00.0000000')
```

I would expect it instead to have generated a where clause using parameters rather than literals.
Edited to paste in actually generated SQL.

**Comment by mburbea** [comment]
Also of ntoe that if the `DateTime2` property was actually a `DateTime` it would still try to send `'2019-07-02T00:00:00.000000'` to the column which would lead to this error: 

> Conversion failed when converting date and/or time from character string.

As Sql server's `DateTime` only support miliseconds, e.g. (3 digits of sub-second precision.)

**Comment by mburbea** [comment]
Still broken in the latest version. This is problematic as it prevents sql server from caching plan shapes. It also prevent usage of legacy types like `DateTime` and `SmallDateTime` which do not support this transmission format.

**Comment by bdebaere** [comment]
> Also of ntoe that if the DateTime2 property was actually a DateTime it would still try to send '2019-07-02T00:00:00.000000' to the column which would lead to this error:
> 
>     > Conversion failed when converting date and/or time from character string.
> 
> As Sql server's DateTime only support miliseconds, e.g. (3 digits of sub-second precision.)

> Still broken in the latest version. This is problematic as it prevents sql server from caching plan shapes. It also prevent usage of legacy types like DateTime and SmallDateTime which do not support this transmission format.

@mburbea This is EntityFrameworkCore's doing. Use the TypeName of the ColumnAttribute to specify the type, like below. That should solve it. In general, it's is a good idea to add the TypeName, because of index usage optimization. For example without it string values are passed as NVARCHAR instead of whatever type you specify in the TypeName of the ColumnAttribute, potentially leading to IMPLICIT_CONVERSION in the query execution plan.

```csharp
public class Issue1843
{
    [Key]
    public MyEnum Enum { get; set; }

    [ColumnAttribute(TypeName = "datetime")] // Alternative example: datetime(2)
    public DateTime DateTime2{ get; set; }
}
```

**Comment by mburbea** [comment]
@bdebaere , this only happens with Odata. When I write a query like so::
```cs
var d=new DateTime(2019,01,01);
return _context.Issue1843.Where(x=> x.DateTime2 >= d);
```
The query generated uses a variable not a literal string.

**Comment by bdebaere** [comment]
@mburbea I wasn't arguing that OData shouldn't pass a parameter. You still have a good point there!
But it is only because T-SQL provides an implicit conversion that this works with a parameter. It is definitely still a good idea to use TypeName of the ColumnAttribute, should this behavior ever change. Consider the T-SQL below.

```tsql
CREATE TABLE Issue1843
(
      Enum INT not null,
      [DateTime2] DATETIME not null
)

DECLARE @parameter DATETIME2(2) = '2019-07-02T00:00:00.0000000'

SELECT * FROM Issue1843 WHERE [DateTime2] = @parameter -- This will succeed due to implicit conversion.
SELECT * FROM Issue1843 WHERE [DateTime2] = '2019-07-02T00:00:00.0000000' -- This will fail.

```

**Comment by ErikNell** [comment]
Using the TypeName property of the ColumnAttribute used to work for us as a workaround. But after upgrading from version 7.1.0 to 7.2.1 this stopped working.

**Comment by KenitoInc** [comment]
This issue was fixed by #2515  in this [commit](https://github.com/OData/WebApi/commit/3a9f4bcc46669ee485d6f576b10b71d94aaec55d)
